### PR TITLE
Fix issue #2446: rotozoom keeps the colorkey flag.

### LINKED ISSUE #2446: pygame.transform.rotozoom loses all surface flags
**Environment:**

pygame-ce 2.3.2 (SDL 2.26.5, Python 3.11.4) on Windows 11

**Current behavior:**

`pygame.transform.rotozoom` loses all surface flags, incl. colorkey.

Docs say "The resulting Surface will be a filtered 32-bit Surface.", but from that I don't read that the "features" of the surface will be lost.

**Expected behavior:**

Returning a surface that preserves surface flags, as all other transforms do (see test script below)

**Screenshots**

Screenshot and output of test script below

image

```py
pygame-ce 2.3.2 (SDL 2.26.5, Python 3.11.4)
original <Surface(64x64x32, colorkey=(0, 0, 0, 255))> 0x1000
rotated <Surface(64x64x32, colorkey=(0, 0, 0, 255))> 0x3000
scaled <Surface(64x64x32, colorkey=(0, 0, 0, 255))> 0x3000
smoothscale <Surface(64x64x32, colorkey=(0, 0, 0, 255))> 0x3000
scaled_by <Surface(64x64x32, colorkey=(0, 0, 0, 255))> 0x3000
smoothscale_by <Surface(64x64x32, colorkey=(0, 0, 0, 255))> 0x3000
rotozoom <Surface(64x64x32)> 0x0
```

**Test code**

If possible add a simple test program that shows the problem described in this report.
```
import pygame

TITLE = 'rotozoom test'
SCREEN = pygame.Rect(0, 0, 1024, 768)
FPS = 60
DT_MAX = 3 / FPS

pygame.init()
pygame.display.set_caption(TITLE)
screen = pygame.display.set_mode(SCREEN.size)
clock = pygame.time.Clock()

image = pygame.Surface((64, 64))
image.set_colorkey('black')
pygame.draw.circle(image, 'red', (32, 32), 32, width=0)

images = {
    'original': image,
    'rotated': pygame.transform.rotate(image, 0),
    'scaled': pygame.transform.scale(image, (64, 64)),
    'smoothscale': pygame.transform.smoothscale(image, (64, 64)),
    'scaled_by': pygame.transform.scale_by(image, 1),
    'smoothscale_by': pygame.transform.smoothscale_by(image, 1),
    'rotozoom': pygame.transform.rotozoom(image, 0, 1),
}

for kind, img in images.items():
    print(kind, img, hex(img.get_flags()))

running = True
while running:
    dt = min(clock.tick(FPS) / 1000.0, DT_MAX)

    for e in pygame.event.get():
        match e.type:
            case pygame.QUIT:
                running = False

            case pygame.KEYDOWN if e.key == pygame.K_ESCAPE:
                running = False

    screen.fill('green')
    x = 50
    for i in images:
        screen.blit(images[i], (x, 50))
        x += 100

    pygame.display.flip()

pygame.quit()
```

### PR DESCRIPTION
**Pull request by igordsm**
Issue #2446 reported that using rotozoom looses the colorkey information and provided code to reproduce the issue. We modified `rotozoomSurface` to copy colorkey from `src` to `rz_dst` and added a test to `test/transform_test.py`.

Closes #2446 

**Before**:
image

**After**:

image

### CODE REVIEW

**Review by ankith26 — CHANGES_REQUESTED** [review]
Hello, thanks for contributing to pygame-ce! :tada: :cake: 

This seems like a nice fix, but we have CI failing on lint errors. To fix that run `setup.py format` and commit resulting changes (make sure you have the appropriate tools like `black`/`clang-format` installed)

File: src_c/transform.c [hunk 1]
```diff
@@ -890,6 +890,9 @@ surf_rotozoom(PyObject *self, PyObject *args, PyObject *kwargs)
 
     Py_BEGIN_ALLOW_THREADS;
     newsurf = rotozoomSurface(surf32, angle, scale, 1);
+    if (newsurf == NULL) {
+        PyErr_SetString(pgExc_SDLError, SDL_GetError());
+    }
```
Comment (ankith26) [inline src_c/transform.c:895]: This block of code should be outside the thread macros. This is because we cannot call python API when the GIL is released. Also, after setting the error, the function must return `NULL`

File: src_c/transform.c [hunk 2]
```diff
@@ -1952,86 +1956,79 @@ clamp_4
...
+            break;                                                            \
+        case 3:                                                               \
+            p_pix = ((Uint8 *)(p_pixels + (p_y)*p_surf->pitch) + (p_x)*3);    \
+            p_color = (SDL_BYTEORDER == SDL_LIL_ENDIAN)                       \
+                          ? (p_pix[0]) + (p_pix[1] << 8) + (p_pix[2] << 16)   \
+                          : (p_pix[2]) + (p_pix[1] << 8) + (p_pix[0] << 16);  \
+            break;                                                            \
+        default: /* case 4: */                                                \
+            p_color = *((Uint32 *)(p_pixels + (p_y)*p_surf->pitch) + (p_x));  \
+            break;                                                            \
```
Comment (MyreMylar) [inline src_c/transform.c:1977]:
```suggestion
#define SURF_GET_AT(p_color, p_surf, p_x, p_y, p_pixels, p_format, p_pix)    \
    switch (p_format->BytesPerPixel) {                                       \
        case 1:                                                              \
            p_color = (Uint32) *                                             \
                      ((Uint8 *)(p_pixels) + (p_y) * p_surf->pitch + (p_x)); \
            break;                                                           \
        case 2:                                                              \
            p_color =                                                        \
                (Uint32) *                                                   \
                ((Uint16 *)((p_pixels) + (p_y) * p_surf->pitch) + (p_x));    \
            break;                                                           \
        case 3:                                                              \
            p_pix =                                                          \
                ((Uint8 *)(p_pixels + (p_y) * p_surf->pitch) + (p_x) * 3);   \
            p_color = (SDL_BYTEORDER == SDL_LIL_ENDIAN)                      \
                          ? (p_pix[0]) + (p_pix[1] << 8) + (p_pix[2] << 16)  \
                          : (p_pix[2]) + (p_pix[1] << 8) + (p_pix[0] << 16); \
            break;                                                           \
        default: /* case 4: */                                               \
            p_color =                                                        \
                *((Uint32 *)(p_pixels + (p_y) * p_surf->pitch) + (p_x));     \
            break;                                                           \
```

File: src_c/transform.c [hunk 3]
```diff
@@ -1952,86 +1956,79 @@ clamp_4
...
+            *(p_byte_buf + (p_format->Rshift >> 3)) =                         \
+                (Uint8)(p_color >> p_format->Rshift);                         \
+            *(p_byte_buf + (p_format->Gshift >> 3)) =                         \
+                (Uint8)(p_color >> p_format->Gshift);                         \
+            *(p_byte_buf + (p_format->Bshift >> 3)) =                         \
+                (Uint8)(p_color >> p_format->Bshift);                         \
+            break;                                                            \
+        default:                                                              \
+            *((Uint32 *)(p_pixels + (p_y)*p_surf->pitch) + (p_x)) = p_color;  \
+            break;                                                            \
```
Comment (MyreMylar) [inline src_c/transform.c:2004]:
```suggestion
#define SURF_SET_AT(p_color, p_surf, p_x, p_y, p_pixels, p_format,       \
                    p_byte_buf)                                          \
    switch (p_format->BytesPerPixel) {                                   \
        case 1:                                                          \
            *((Uint8 *)p_pixels + (p_y) * p_surf->pitch + (p_x)) =       \
                (Uint8)p_color;                                          \
            break;                                                       \
        case 2:                                                          \
            *((Uint16 *)(p_pixels + (p_y) * p_surf->pitch) + (p_x)) =    \
                (Uint16)p_color;                                         \
            break;                                                       \
        case 3:                                                          \
            p_byte_buf =                                                 \
                (Uint8 *)(p_pixels + (p_y) * p_surf->pitch) + (p_x) * 3; \
            *(p_byte_buf + (p_format->Rshift >> 3)) =                    \
                (Uint8)(p_color >> p_format->Rshift);                    \
            *(p_byte_buf + (p_format->Gshift >> 3)) =                    \
                (Uint8)(p_color >> p_format->Gshift);                    \
            *(p_byte_buf + (p_format->Bshift >> 3)) =                    \
                (Uint8)(p_color >> p_format->Bshift);                    \
            break;                                                       \
        default:                                                         \
            *((Uint32 *)(p_pixels + (p_y) * p_surf->pitch) + (p_x)) =    \
                p_color;                                                 \
            break;                                                       \
```

File: src_c/transform.c [hunk 4]
```diff
@@ -1952,86 +1956,79 @@ clamp_4
...
+            *(p_byte_buf + 2 - (p_format->Rshift >> 3)) =                     \
+                (Uint8)(p_color >> p_format->Rshift);                         \
+            *(p_byte_buf + 2 - (p_format->Gshift >> 3)) =                     \
+                (Uint8)(p_color >> p_format->Gshift);                         \
+            *(p_byte_buf + 2 - (p_format->Bshift >> 3)) =                     \
+                (Uint8)(p_color >> p_format->Bshift);                         \
+            break;                                                            \
+        default:                                                              \
+            *((Uint32 *)(p_pixels + (p_y)*p_surf->pitch) + (p_x)) = p_color;  \
+            break;                                                            \
```
Comment (MyreMylar) [inline src_c/transform.c:2031]:
```suggestion
#define SURF_SET_AT(p_color, p_surf, p_x, p_y, p_pixels, p_format,       \
                    p_byte_buf)                                          \
    switch (p_format->BytesPerPixel) {                                   \
        case 1:                                                          \
            *((Uint8 *)p_pixels + (p_y) * p_surf->pitch + (p_x)) =       \
                (Uint8)p_color;                                          \
            break;                                                       \
        case 2:                                                          \
            *((Uint16 *)(p_pixels + (p_y) * p_surf->pitch) + (p_x)) =    \
                (Uint16)p_color;                                         \
            break;                                                       \
        case 3:                                                          \
            p_byte_buf =                                                 \
                (Uint8 *)(p_pixels + (p_y) * p_surf->pitch) + (p_x) * 3; \
            *(p_byte_buf + 2 - (p_format->Rshift >> 3)) =                \
                (Uint8)(p_color >> p_format->Rshift);                    \
            *(p_byte_buf + 2 - (p_format->Gshift >> 3)) =                \
                (Uint8)(p_color >> p_format->Gshift);                    \
            *(p_byte_buf + 2 - (p_format->Bshift >> 3)) =                \
                (Uint8)(p_color >> p_format->Bshift);                    \
            break;                                                       \
        default:                                                         \
            *((Uint32 *)(p_pixels + (p_y) * p_surf->pitch) + (p_x)) =    \
                p_color;                                                 \
            break;                                                       \
```

File: src_c/rotozoom.c [hunk 5]
```diff
@@ -583,6 +584,13 @@ rotozoomSurface(SDL_Surface *src, double angle, double zoom, int smooth)
          * Target surface is 32bit with source RGBA/ABGR ordering
          */
         rz_dst = PG_CreateSurface(dstwidth, dstheight, rz_src->format->format);
+        if (SDL_GetColorKey(src, &colorkey) == 0) {
+            if (SDL_SetColorKey(rz_dst, SDL_TRUE, colorkey) != 0 ||
+                SDL_SetSurfaceRLE(rz_dst, SDL_TRUE) != 0) {
```
Comment (MyreMylar) [inline src_c/rotozoom.c:589]: IMHO we shouldn't be using this pattern in `transform.c` of always enabling RLE whenever a source surface has a color key. RLE no longer always works in all circumstances.

See this issue: https://github.com/pygame-community/pygame-ce/issues/1668

Instead I think we should check the source surface to see if it already has RLE, and if so, then enable it for the destination. That should isolate any RLE & source alpha bugs to when we are initially creating surfaces and when calling `Surface.set_alpha()`.

Slightly complicating this, is that the function to get a surface's RLE is one of the one's being renamed in SDL3. It is currently called: `SDL_HasSurfaceRLE()` but is going to be called `SDL_SurfaceHasRLE()` in SDL3.

In short I think we need to make something like what this PR has added for `SDL_ConvertSurface()` in `_pygame.h` but with `PG_SurfaceHasRLE()` as the name and then use it here something like:

```C
    if (SDL_GetColorKey(src, &colorkey) == 0) {
        if (SDL_SetColorKey(rz_dst, SDL_TRUE, colorkey) != 0) {
            SDL_FreeSurface(rz_dst);
            return NULL;
        }
        if(PG_SurfaceHasRLE(src) && SDL_SetSurfaceRLE(rz_dst, SDL_TRUE) != 0) {
            SDL_FreeSurface(rz_dst);
            return NULL;
        }
    }
```

**Review by MyreMylar — CHANGES_REQUESTED** [review]
Just added a note here about an RLE thing that likely needs fixing in general across the whole of `transform.c` before we add another version of it in here.

We probably also need to make some tests for it based on the first post in #1668 just to check if it still is, or isn't a problem, as SDL has likely changed since this issue was first posted.

**Comment by MyreMylar** [comment]
This will also need to merge in the latest main branch to pick up the grub-efi fix.

File: src_c/rotozoom.c [hunk 6]
```diff
@@ -628,7 +640,18 @@ rotozoomSurface(SDL_Surface *src, double angle, double zoom, int smooth)
...
          * Target surface is 32bit with source RGBA/ABGR ordering
          */
+
         rz_dst = PG_CreateSurface(dstwidth, dstheight, rz_src->format->format);
+        if (SDL_GetColorKey(src, &colorkey) == 0) {
+            if (SDL_SetColorKey(rz_dst, SDL_TRUE, colorkey) != 0) {
+                SDL_FreeSurface(rz_dst);
+                return NULL;
+            }
+            if (PG_SurfaceHasRLE(src) && SDL_SetSurfaceRLE(rz_dst, SDL_TRUE) != 0) {
```
Comment (MyreMylar) [inline src_c/rotozoom.c:650]:
```suggestion
            if (PG_SurfaceHasRLE(src) &&
                SDL_SetSurfaceRLE(rz_dst, SDL_TRUE) != 0) {
```

File: src_c/rotozoom.c [hunk 7]
```diff
@@ -583,6 +585,16 @@ rotozoomSurface(SDL_Surface *src, double angle, double zoom, int smooth)
          * Target surface is 32bit with source RGBA/ABGR ordering
          */
         rz_dst = PG_CreateSurface(dstwidth, dstheight, rz_src->format->format);
+        if (SDL_GetColorKey(src, &colorkey) == 0) {
+            if (SDL_SetColorKey(rz_dst, SDL_TRUE, colorkey) != 0) {
+                SDL_FreeSurface(rz_dst);
+                return NULL;
+            }
+            if (PG_SurfaceHasRLE(src) && SDL_SetSurfaceRLE(rz_dst, SDL_TRUE) != 0) {
```
Comment (MyreMylar) [inline src_c/rotozoom.c:593]:
```suggestion
            if (PG_SurfaceHasRLE(src) &&
                SDL_SetSurfaceRLE(rz_dst, SDL_TRUE) != 0) {
```

**Review by MyreMylar — APPROVED** [review]
LGTM. Can confirm it fixes the issue and works with RLE and non-RLE colorkey surfaces. Thank you for the contribution!

👍 🎉

File: src_c/_pygame.h [hunk 8]
```diff
@@ -72,6 +72,8 @@
 #define PG_ConvertSurface SDL_ConvertSurface
 #define PG_ConvertSurfaceFormat SDL_ConvertSurfaceFormat
 
+#define PG_SurfaceHasRLE(src) SDL_SurfaceHasRLE(src)
```
Comment (MyreMylar) [inline src_c/_pygame.h:75]:
```suggestion
#define PG_SurfaceHasRLE SDL_SurfaceHasRLE
```

File: src_c/_pygame.h [hunk 9]
```diff
@@ -100,7 +102,7 @@
 #define PG_ConvertSurface(src, fmt) SDL_ConvertSurface(src, fmt, 0)
 #define PG_ConvertSurfaceFormat(src, pixel_format) \
     SDL_ConvertSurfaceFormat(src, pixel_format, 0)
-
+#define PG_SurfaceHasRLE(src) SDL_HasSurfaceRLE(src)
```
Comment (MyreMylar) [inline src_c/_pygame.h:105]:
```suggestion
#define PG_SurfaceHasRLE SDL_HasSurfaceRLE
```

File: src_c/_pygame.h [hunk 10]
```diff
@@ -100,7 +102,7 @@
 #define PG_ConvertSurface(src, fmt) SDL_ConvertSurface(src, fmt, 0)
 #define PG_ConvertSurfaceFormat(src, pixel_format) \
     SDL_ConvertSurfaceFormat(src, pixel_format, 0)
-
+#define PG_SurfaceHasRLE SDL_HasSurfaceRLE
```
Comment (MyreMylar) [inline src_c/_pygame.h:105]:
```suggestion
#if SDL_VERSION_ATLEAST(2, 0, 14)
#define PG_SurfaceHasRLE SDL_HasSurfaceRLE
#else
// vendored in until our lowest SDL version is 2.0.14
SDL_bool
PG_SurfaceHasRLE(SDL_Surface *surface)
{
    if (surface == NULL) {
        return SDL_FALSE;
    }

    if (!(surface->map->info.flags & SDL_COPY_RLE_DESIRED)) {
        return SDL_FALSE;
    }

    return SDL_TRUE;
}
#endif

```

File: src_c/_pygame.h [hunk 11]
```diff
@@ -100,6 +102,24 @@
 #define PG_ConvertSurface(src, fmt) SDL_ConvertSurface(src, fmt, 0)
 #define PG_ConvertSurfaceFormat(src, pixel_format) \
     SDL_ConvertSurfaceFormat(src, pixel_format, 0)
+#if SDL_VERSION_ATLEAST(2, 0, 14)
+#define PG_SurfaceHasRLE SDL_HasSurfaceRLE
+#else
+// vendored in until our lowest SDL version is 2.0.14
```
Comment (MyreMylar) [inline src_c/_pygame.h:108]:
```suggestion
// vendored in until our lowest SDL version is 2.0.14
#include "_blit_info.h"
#define SDL_COPY_RLE_DESIRED 0x00001000

```

**Comment by MyreMylar** [comment]
I realised while investigating the failing test that the function now known as `SDL_SurfaceHasRLE` that was renamed for SDL 3, was also only added to SDL 2 in version 2.0.14 so we are still 5 patch versions behind in our lowest supported SDL to use it. notably our wheels are ahead of this version so that's why I wasn't seeing this error locally.

I've just added a patch commit to vendor this function in for the case where SDL is less that 2.0.14. Not the most elegant solution (if it eventually works), but it should hold up OK as the function itself is super basic and we can delete it eventually when our bottom end catches up with our current  main target.

File: src_c/_pygame.h [hunk 12]
```diff
@@ -100,6 +102,27 @@
...
+#include "_blit_info.h"
```
Comment (MyreMylar) [inline src_c/_pygame.h:109]:
```suggestion
struct SDL_BlitMap
{
    SDL_Surface *dst;
    int identity;
    SDL_blit blit;
    void *data;
    SDL_BlitInfo info;

    /* the version count matches the destination; mismatch indicates
       an invalid mapping */
    Uint32 dst_palette_version;
    Uint32 src_palette_version;
};
```

File: src_c/_pygame.h [hunk 13]
```diff
@@ -100,6 +102,39 @@
...
+    int identity;
+    SDL_blit blit;
+    void *data;
+    SDL_BlitInfo info;
+
+    /* the version count matches the destination; mismatch indicates
+       an invalid mapping */
+    Uint32 dst_palette_version;
+    Uint32 src_palette_version;
+};
```
Comment (MyreMylar) [inline src_c/_pygame.h:121]:
```suggestion
typedef struct {
    int width;
    int height;
    Uint8 *s_pixels;
    int s_pxskip;
    int s_skip;
    Uint8 *d_pixels;
    int d_pxskip;
    int d_skip;
    SDL_PixelFormat *src;
    SDL_PixelFormat *dst;
    Uint8 src_blanket_alpha;
    int src_has_colorkey;
    Uint32 src_colorkey;
    SDL_BlendMode src_blend;
    SDL_BlendMode dst_blend;
} SDL_BlitInfo;

struct SDL_BlitMap {
    SDL_Surface *dst;
    int identity;
    SDL_blit blit;
    void *data;
    SDL_BlitInfo info;

    /* the version count matches the destination; mismatch indicates
       an invalid mapping */
    Uint32 dst_palette_version;
    Uint32 src_palette_version;
};
```

File: src_c/rotozoom.c [hunk 14]
```diff
@@ -11,6 +11,7 @@
 
 */
 
+#include "_pygame.h"
```
Comment (MyreMylar) [inline src_c/rotozoom.c:14]:
```suggestion
#include "_pygame.h"

#if !SDL_VERSION_ATLEAST(2, 0, 14)
SDL_bool
PG_SurfaceHasRLE(SDL_Surface *surface)
{
    if (surface == NULL) {
        return SDL_FALSE;
    }

    if (!(surface->map->info.flags & SDL_COPY_RLE_DESIRED)) {
        return SDL_FALSE;
    }

    return SDL_TRUE;
}
#endif
```

**Comment by MyreMylar** [comment]
Getting that to compile via Github web version interactions required a few compromises... I'm planning to do another PR on the topic of Surface RLE flags for the whole transform module once this gets merged, so I can almost certainly clean this up a bit in that PR.

File: src_c/_pygame.h [hunk 15]
```diff
@@ -72,6 +72,8 @@
 #define PG_ConvertSurface SDL_ConvertSurface
 #define PG_ConvertSurfaceFormat SDL_ConvertSurfaceFormat
 
+#define PG_SurfaceHasRLE SDL_SurfaceHasRLE
```
Comment (ankith26) [inline src_c/_pygame.h:75]: This line should be removed?

File: src_c/_pygame.h [hunk 16]
```diff
@@ -100,6 +102,44 @@
...
+    int identity;
+    SDL_blit blit;
+    void *data;
+    SDL_InternalBlitInfo info;
+
+    /* the version count matches the destination; mismatch indicates
+       an invalid mapping */
+    Uint32 dst_palette_version;
+    Uint32 src_palette_version;
+};
```
Comment (ankith26) [inline src_c/_pygame.h:137]: I don't why these two structs are being vendored in? Just vendoring in `SDL_COPY_RLE_DESIRED` should be enough, unless I'm missing something

File: src_c/_pygame.h (hunk 15 above)
Comment (MyreMylar) [reply src_c/_pygame.h:75]: No, that is there for SDL 3 compatibility as they have renamed `SDL_HasSurfaceRLE` to `SDL_SurfaceHasRLE`.

File: src_c/_pygame.h (hunk 16 above)
Comment (MyreMylar) [reply src_c/_pygame.h:137]: I tried not including the structs (you can see the series of commits) the compiler that goes down this route using SDL 2.0.10 errors on line 24 below when getting the pointer to `map` unless both of these structs are defined. 

They are not public structs as they are part of the SDL internals so to use them at all we have to vendor them in temporarily as well. The versions of SDL this applies to (2.0.9 to 2.0.13) are already released so their struct internals won't be changing so I don't see any long term harm from having them here until it comes time to remove them when the bottom end of our SDL support moves up.

**Review by ankith26 — APPROVED** [review]
LGTM, thanks for the PR :tada:

**Comment by ankith26** [comment]
Noticed the original commit has many co-authors, so thank you all! Unless anyone minds, I'd like to squash-and-merge this PR but I will leave the co-author tags intact.

**Comment by MyreMylar** [comment]
Absolutely squash and merge away. Igor did 99% of the work here. I was
just wrestling with the compiler to get it over the line.
